datafusion/functions-nested/src/range.rs

```diff
@@ -394,3 +412,136 @@ fn gen_range_date(args: &[ArrayRef], include_upper: bool) -> Result<ArrayRef> {
 
     Ok(arr)
 }
+
+fn gen_range_timestamp(args: &[ArrayRef], include_upper: bool) -> Result<ArrayRef> {
+    if args.len() != 3 {
+        return exec_err!(
+            "arguments length must be 3 for {}",
+            if include_upper {
+                "generate_series"
+            } else {
+                "range"
+            }
+        );
+    }
+
+    // coerce_types fn should coerce all types to Timestamp(Nanosecond, tz)
+    let (start_arr, start_tz_opt) = cast_timestamp_arg(&args[0], include_upper)?;
+    let (stop_arr, stop_tz_opt) = cast_timestamp_arg(&args[1], include_upper)?;
+    let step_arr = as_interval_mdn_array(&args[2])?;
+    let start_tz = parse_tz(start_tz_opt)?;
+    let stop_tz = parse_tz(stop_tz_opt)?;
+
+    // values are timestamps
+    let values_builder = if let Some(start_tz_str) = start_tz_opt {
+        TimestampNanosecondBuilder::new().with_timezone(&**start_tz_str)
```

Comment:
that looks somewhat weird, do we really need so many indirections here?

Reply:
As it was written, yes, it was required. I changed it up to use map_or_else to accomplish the same thing.